internal/middleware/etag_middleware.go

```diff
@@ -0,0 +1,39 @@
+package middleware
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+)
+
+var InvalidETagErr = errors.New("empty etag provided")
+
+func ETagMiddleware(expires time.Duration, name, etag string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			if etag == "" {
+				panic(InvalidETagErr)
+			}
+			etagValue := fmt.Sprintf("\"hms-%s-%s\"", name, etag)
+			controlValue := fmt.Sprintf("max-age=%d", int(expires.Seconds()))
+			w.Header().Set("ETag", etagValue)
+			w.Header().Set("Cache-Control", controlValue)
+			logger := ctxval.Logger(r.Context()).With().Str("etag", etag).Str("etag_name", name).Logger()
+			logger.Trace().Msgf("Returned etag with Cache-Control '%s'", controlValue)
+
+			if match := r.Header.Get("If-None-Match"); match != "" {
+				if strings.Contains(match, etag) {
+					logger.Trace().Msgf("ETag cache hit")
+					w.WriteHeader(http.StatusNotModified)
+					return
+				}
+			}
+			next.ServeHTTP(w, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
```

Comment:
The middleware responsible for setting the ETag and Cache-Control and respecting the If-None-Match header from the client to quickly return NotModified if browser already has that specific version.